sklearn/inspection/_h_statistic.py

```diff
@@ -0,0 +1,260 @@
+"""Friedman and Popescu's H-Statistic"""
+
+import itertools
+
+import numpy as np
+
+from ..base import is_classifier, is_regressor
+from ..utils import Bunch, _get_column_indices, _safe_assign, _safe_indexing
+from ..utils.random import sample_without_replacement
+from ..utils.validation import _check_sample_weight, check_is_fitted
+
+
+def _calculate_pd_brute_fast(estimator, X, feature_indices, grid, sample_weight=None):
+    """Fast version of _calculate_partial_dependence_brute()
+
+    Returns np.array of size (n_grid, ) or (n_ngrid, output_dimension).
+    """
+
+    if is_regressor(estimator):
+        pred_fun = estimator.predict
+    elif is_classifier(estimator) and hasattr(estimator, "predict_proba"):
+        pred_fun = estimator.predict_proba
+    else:
+        raise ValueError("The estimator has no predict or predict_proba method.")
+
+    # X is stacked n_grid times, and grid columns are replaced by replicated grid
+    n = X.shape[0]
+    n_grid = grid.shape[0]
+    X_eval = X.copy()
+
+    X_stacked = _safe_indexing(X_eval, np.tile(np.arange(n), n_grid), axis=0)
+    grid_stacked = _safe_indexing(grid, np.repeat(np.arange(n_grid), n), axis=0)
+    _safe_assign(X_stacked, values=grid_stacked, column_indexer=feature_indices)
+
+    # Predict on stacked data. Pick positive class probs for binary classification
+    preds = pred_fun(X_stacked)
+    if is_classifier(estimator) and preds.shape[1] == 2:
+        preds = preds[:, 1]
+
+    # Partial dependences are averages per grid block
+    pd_values = [
+        np.average(Z, axis=0, weights=sample_weight) for Z in np.split(preds, n_grid)
+    ]
+
+    return np.array(pd_values)
+
+
+def _calculate_pd_over_data(estimator, X, feature_indices, sample_weight=None):
+    """Calculates centered partial dependence over the data distribution.
+
+    It returns a numpy array of size (n, ) or (n, output_dimension).
```

Comment:
What is output_dimension?

Reply:
It is the number of values returned by predicting a single observation, with the exception of binary classification (probabilistic) where it is 1. Do we need to say this somewhere?

Reply:
It would not hurt.